### PR TITLE
Slate/HTML serializers and deserializers

### PR DESCRIPTION
**Pull request by tiberiuichim**
The html2slate and slate2html are the value here.

For the curious, there's an `htmlblock.zcml` that you can include, which will override the way slate blocks are stored, making them always store as HTML in ZODB, with python-side serializer/deserializer.

Checklist:

- [x] Solve resiliparse dependency problem
- [x] Documentation
- [x] Rewrite the link support to the model is used in Volto, it's the old EEA style links

### CODE REVIEW

**Comment by tiberiuichim** [comment]
@jenkins-plone-org please run jobs

**Comment by tiberiuichim** [comment]
Jenkins is complaining that I've added a new dependency and it's not pinned. 

What's the procedure for this?

image

[Resiliparser](https://github.com/chatnoir-eu/chatnoir-resiliparse) is the only Python HTML parser that represents its data as DOM Nodes (Document, Element, Inlines and Text nodes) and so it's the only one that I could use in a convenient manner to preserve the perception-based rendering of HTML in browser. When I've initially created eea.volto.slate back in 2021, this parser didn't exist (and I was an ignorant back then to all the complexities of this problem space). So, I'm ready to be pointed to another option, but only if it can do browser-style dom parsing. html5lib won't cut it. It doesn't expose TextNodes, all text between tags is exposed as node.tail. So, a wrapper could be done on top of it, but it's more work then it's worth.

**Comment by davisagli** [comment]
@tiberiuichim I'm not familiar with the motivation for this work, so can you say a bit more about why you want to store blocks as HTML? My initial gut reaction is that it sounds like it adds unnecessary work at the time of serialization and deserialization, and also adds complexity. Is there a benefit that makes that worthwhile?

Regarding Resiliparse, I would want to avoid adding a new dependency if possible. Did you consider BeautifulSoup (bs4), which is already a dependency of Plone? It looks like it exposes each bit of text as a https://www.crummy.com/software/BeautifulSoup/bs4/doc/#navigablestring -- but maybe there is some other thing it doesn't handle in the way you need?

**Comment by tiberiuichim** [comment]
@davisagli I didn't know about the BS4 text nodes (I tend to use lxml + lxml.html.cleaner), I'll see if I can make them fit. Thanks for the pointer!

The slate html block transformer is a "tech demo". It is a way to test that the serializer/deserializers work correctly, as I can use them directly with the text block in Volto. That's why the whole `htmlblock.zcml` is optional, it's not included by default.

With that said, I would argue this: storing HTML is a lot safer, long term, then storing our custom flavour of JSON that represents the Slate trees. HTML is complicated and it's not pretty, but you can't compare the amount of support that it has. Our Slate JSON support is just two libraries: the nodejs block converter and this code. Perhaps we could store also the HTML value in the DX blocks field, for "long term storage".

And you can't escape dealing with HTML. One task we'll have to approach next year is automatically translating Plone content (with Volto blocks) using the eTranslation service. It can ingest HTML and return translated HTML. So the html => slate and slate => html is a must.

**Comment by tiberiuichim** [comment]
On the whole motivation of work, right now the Plone 6 migration needs to interface with the block converter, complicating the overall procedure. With this, it can be done completely from Python

**Comment by davisagli** [comment]
@tiberiuichim Thanks, that's helpful. I can definitely see some potential uses of being able to do this transformation in Python as well as Node. I'll try to take a closer look soon.

I do wonder if it should go in a separate library, so that it could be used by a Python client of the REST API without depending on all of Plone. On the other hand we have too many Python packages in Plone already...

**Comment by tiberiuichim** [comment]
> I do wonder if it should go in a separate library, so that it could be used by a Python client of the REST API without depending on all of Plone. 

The [html2slate.py](https://github.com/plone/plone.volto/blob/039354dfac505e923c42c7314292e34525e017c3/src/plone/volto/slate/html2slate.py) and [slate2html.py](https://github.com/plone/plone.volto/blob/039354dfac505e923c42c7314292e34525e017c3/src/plone/volto/slate/slate2html.py) are not dependent on any Plone. If anyone needs them in their own projects, they can just copy them.

**Comment by tiberiuichim** [comment]
@jenkins-plone-org please run jobs

**Comment by tiberiuichim** [comment]
I think it's ready.

**Comment by tiberiuichim** [comment]
@jenkins-plone-org please run jobs

File: src/plone/volto/tests/data/6-1.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<ul>
```
Comment (davisagli) [inline src/plone/volto/tests/data/6-1.html:1]: Why is there 6-1.html and also 6.html?

File: src/plone/volto/tests/test_html2slate.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+        )
+
+    def test_ol_sample(self):
+        """Encountered while testing"""
+        text = read_data("9.html")
+        res = text_to_slate(text)
+        self.assertEqual(
+            res,
+            read_json("9.json"),
+        )
```
Comment (davisagli) [inline src/plone/volto/tests/test_html2slate.py:422]: It might make sense to use https://pypi.org/project/parameterized/ for these tests that are basically the same as each other except for the test data.

**Review by davisagli — COMMENTED** [review]
@tiberiuichim Here's my first pass on reviewing. So far I looked at the tests to make sure that what it's doing makes sense, but I didn't yet look at how it's doing it.

File: README.rst [hunk 3]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
 sure your add-on's profiles are applied AFTER this one.**
 
 
+Slate support
+-------------
+
+Starting with Volto 16, the default richtext implementation uses SlateJS. We
+include 2 converters:
+
+- Volto's flavour of Slate JSON to HTML (``Slate2HTML``)
```
Comment (stevepiercy) [inline README.rst:269]: Please use US not UK spellings.

```suggestion
- Volto's flavor of Slate JSON to HTML (``Slate2HTML``)
```

File: README.rst [hunk 4]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+- HTML to Volto's flavour of Slate JSON (``HTML2Slate``)
```
Comment (stevepiercy) [inline README.rst:270]:
```suggestion
- HTML to Volto's flavor of Slate JSON (``HTML2Slate``)
```

File: README.rst [hunk 5]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+
+These two classes can be inherited and extended for your custom tags and
+plugins.
+
+It's also possible to include an optional ``htmlblock.zcml`` configure file
```
Comment (stevepiercy) [inline README.rst:275]:
```suggestion
It's also possible to include an optional ``htmlblock.zcml`` configuration file,
```

File: README.rst [hunk 6]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+(located in ``plone.volto.slate`` ). The effect is that the value of the slate
```
Comment (stevepiercy) [inline README.rst:276]:
```suggestion
located in ``plone.volto.slate``. The effect is that the value of the Slate
```

File: README.rst [hunk 7]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+blocks will be stored as an HTML and it will be transparently converted to
```
Comment (stevepiercy) [inline README.rst:277]:
```suggestion
blocks will be stored as HTML, and it will be transparently converted to
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Minor grammar fixes to README.rst

File: src/plone/volto/tests/data/6-1.html (hunk 1 above)
Comment (tiberiuichim) [reply src/plone/volto/tests/data/6-1.html:1]: Because 6.html is a dirty html, while 6-1.html is clean output produced by serializing Slate JSON, so the expectation is that the deserializer ingests the dirty, but produces clean.

File: src/plone/volto/tests/data/5.json [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    "type": "p",
+    "children": [
+      {
+        "text": "Paragraph "
+      },
+      {
+        "type": "strong",
+        "children": [
+          {
+            "text": ""
```
[the remark this answers is not in the document]
Comment (tiberiuichim) [reply src/plone/volto/tests/data/5.json:12]: Those text nodes are intentional, see https://github.com/plone/plone.volto/blob/9207feb01a71ad62c2b52cd9161dacbfada313df/src/plone/volto/slate/html2slate.py#L385-L412

[the remark this answers is not in the document]
Comment (tiberiuichim) [reply src/plone/volto/tests/data/5.json:12]: @davisagli Slate needs those empty text nodes to know where to place a cursor, when you click at the beginning or end of an "inline block element".

File: src/plone/volto/tests/test_html2slate.py (hunk 2 above)
Comment (tiberiuichim) [reply src/plone/volto/tests/test_html2slate.py:422]: @davisagli Good pointer, I knew about the parametrized tests for pytest, but not for the unittest. I actually briefly looked for a solution while developing the tests, but I felt that it would make the development harder, as I needed to work one test at a time, debug, etc. This would be harder with parametric tests. I think they're not that many anyway (though I hope we add more, if we find bugs).

**Review by stevepiercy — APPROVED** [review]

File: src/plone/volto/tests/data/6-1.html (hunk 1 above)
Comment (davisagli) [reply src/plone/volto/tests/data/6-1.html:1]: I see, it does not roundtrip perfectly because the whitespace is normalized.

File: src/plone/volto/tests/data/5.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (davisagli) [reply src/plone/volto/tests/data/5.json:12]: @tiberiuichim I'm going to take a closer look at this, because it's not what blocks-conversion-tool is doing and I need to understand whether it is producing output that will be a problem.

My gut feeling is that even if these empty text nodes are needed at runtime, they should be added implicitly when the editor loads and filtered out when the data is serialized, rather than being part of the stored data. They have no semantic value.

File: src/plone/volto/tests/test_html2slate.py (hunk 2 above)
Comment (davisagli) [reply src/plone/volto/tests/test_html2slate.py:422]: @tiberiuichim Fair point.

File: src/plone/volto/tests/data/5.json (hunk 8 above)
[the remark this answers is not in the document]
Comment (tiberiuichim) [reply src/plone/volto/tests/data/5.json:12]: @davisagli The intention was to perform a perfect roundtrip. Indeed, a normalization at JS volto-slate level is probably better. 

It is possible that slatejs itself is less sensitive to those missing text nodes in the values, I used to get a lot of crashes when the value that was injected as a prop (I don't remember my exact test scenarios, it's possible that I was trying to push it too much).

[the remark this answers is not in the document]
Comment (tiberiuichim) [reply src/plone/volto/tests/data/5.json:12]: @davisagli I test, the padding is no longer needed. I'll disable it.

[the remark this answers is not in the document]
Comment (davisagli) [reply src/plone/volto/tests/data/5.json:12]: @tiberiuichim nice!

File: README.rst [hunk 9]
```diff
@@ -260,6 +260,23 @@ existing ones. These are know to work well with Volto layout and grid system::
...
+Slate, when the content is loaded in Volto.
```
[the remark this answers is not in the document]
Comment (tiberiuichim) [reply README.rst:278]: If we have perfect converters in/out, I think it's better to store html instead of custom JSON. But this deserves a discussion and a strategy on how to expose this to other Plone projects.

For the moment I'll add the note you suggest.

**Comment by tiberiuichim** [comment]
@jenkins-plone-org please run jobs

File: src/plone/volto/slate/html2slate.py [hunk 10]
```diff
@@ -0,0 +1,398 @@
...
+        return element
+
+    def handle_fallback(self, node):
+        """Unknown tags (for example span) are handled as pipe-through"""
+        return self.deserialize_children(node)
+
+    def normalize(self, value):
+        """Normalize value to match Slate constraints"""
+
+        assert isinstance(value, list)
```
Comment (davisagli) [inline src/plone/volto/slate/html2slate.py:360]: If this isn't the case, the program will end with AssertionError: False which won't be super helpful. If this is something that should never happen, using type annotations and static analysis in your editor might be a better way to go. If we expect that we might actually get a list here from certain user input, then we should give the user a more helpful error about what they did wrong.

**Review by davisagli — COMMENTED** [review]
@tiberiuichim Sorry for taking so long to get back to this. I've now done a pass reading through the implementation. For something like this it's hard to say when enough testing has been done, since there are always edge cases. It would be possible to go borrow some more test cases from blocks-conversion-tool. But, maybe it makes as much sense to merge it as is and let someone try it with real data.
